File: SyncFlowMac/SyncFlowMac/Services/ClipboardSyncService.swift
//
//  ClipboardSyncService.swift
//  SyncFlowMac
//
//  Service to sync clipboard content between macOS and Android
//

import Foundation
import AppKit
import FirebaseDatabase
import Combine

class ClipboardSyncService: ObservableObject {
    static let shared = ClipboardSyncService()

    @Published var isEnabled: Bool = true
    @Published var lastSyncedContent: String?
    @Published var lastSyncTime: Date?

    private let database = Database.database()
    private var clipboardHandle: DatabaseHandle?
    private var currentUserId: String?
    private var clipboardCheckTimer: Timer?

    // Track last synced content to avoid loops
    private var lastKnownContent: String?
    private var lastKnownChangeCount: Int = 0
    private var isUpdatingFromRemote = false

    private let maxClipboardLength = 50000 // 50KB max

    private init() {}

    /// Start clipboard sync
    func startSync(userId: String) {
        guard isEnabled else { return }
        currentUserId = userId

        // Start monitoring local clipboard
        startMonitoringLocalClipboard()

        // Start listening for remote clipboard changes
        startListeningForRemoteClipboard(userId: userId)

        print("ClipboardSyncService: Started syncing for user \(userId)")
    }

    /// Stop clipboard sync
    func stopSync() {
        stopMonitoringLocalClipboard()
        stopListeningForRemoteClipboard()
        currentUserId = nil
        print("ClipboardSyncService: Stopped syncing")
    }

    /// Monitor local clipboard for changes
    private func startMonitoringLocalClipboard() {
        // macOS doesn't have a notification for clipboard changes
        // So we poll the clipboard periodically
        DispatchQueue.main.async { [weak self] in
            self?.lastKnownChangeCount = NSPasteboard.general.changeCount

            self?.clipboardCheckTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
                self?.checkLocalClipboard()
            }
            // Ensure timer runs on main run loop
            if let timer = self?.clipboardCheckTimer {
                RunLoop.main.add(timer, forMode: .common)
            }
            print("ClipboardSyncService: Local clipboard monitoring started")
        }
    }

    private func stopMonitoringLocalClipboard() {
        clipboardCheckTimer?.invalidate()
        clipboardCheckTimer = nil
    }

    /// Check if local clipboard has changed
    private func checkLocalClipboard() {
        guard isEnabled, !isUpdatingFromRemote else { return }
        guard currentUserId != nil else { return }

        let pasteboard = NSPasteboard.general
        let currentChangeCount = pasteboard.changeCount

        // Check if clipboard changed
        guard currentChangeCount != lastKnownChangeCount else { return }
        lastKnownChangeCount = currentChangeCount

        // Get text content
        guard let text = pasteboard.string(forType: .string) else {
            print("ClipboardSyncService: No text in clipboard")
            return
        }
        guard !text.isEmpty else { return }
        guard text.count <= maxClipboardLength else {
            print("ClipboardSyncService: Content too large (\(text.count) chars), skipping")
            return
        }

        // Check if content actually changed
        guard text != lastKnownContent else { return }

        print("ClipboardSyncService: Local clipboard changed, syncing: \(text.prefix(50))...")
        syncToFirebase(text: text)
    }

    /// Listen for clipboard changes from other devices
    private func startListeningForRemoteClipboard(userId: String) {
        let clipboardRef = database.reference()
            .child("users")
            .child(userId)
            .child("clipboard")

        clipboardHandle = clipboardRef.observe(.value) { [weak self] snapshot in
            guard let self = self else { return }

            guard let data = snapshot.value as? [String: Any] else { return }
            guard let source = data["source"] as? String else { return }

            // Only process if from another device (not macOS)
            guard source != "macos" else { return }

            guard let text = data["text"] as? String else { return }
            guard let timestamp = data["timestamp"] as? Double else { return }

            // Check if this is newer than what we synced
            if let lastSync = self.lastSyncTime,
               Date(timeIntervalSince1970: timestamp / 1000) <= lastSync {
                return
            }

            print("ClipboardSyncService: Received clipboard from \(source)")
            self.updateLocalClipboard(text: text, timestamp: timestamp)
        }
    }

    private func stopListeningForRemoteClipboard() {
        guard let userId = currentUserId, let handle = clipboardHandle else { return }

        database.reference()
            .child("users")
            .child(userId)
            .child("clipboard")
            .removeObserver(withHandle: handle)

        clipboardHandle = nil
    }

    /// Sync local clipboard to Firebase
    private func syncToFirebase(text: String) {
        guard let userId = currentUserId else { return }

        let clipboardRef = database.reference()
            .child("users")
            .child(userId)
            .child("clipboard")

        let timestamp = Date().timeIntervalSince1970 * 1000

        let data: [String: Any] = [
            "text": text,
            "timestamp": timestamp,
            "source": "macos",
            "type": "text"
        ]

        clipboardRef.setValue(data) { [weak self] error, _ in
            if let error = error {
                print("ClipboardSyncService: Error syncing to Firebase: \(error)")
            } else {
                DispatchQueue.main.async {
                    self?.lastKnownContent = text
                    self?.lastSyncedContent = text
                    self?.lastSyncTime = Date()
                }
                print("ClipboardSyncService: Synced to Firebase")
            }
        }
    }

    /// Update local clipboard with remote content
    private func updateLocalClipboard(text: String, timestamp: Double) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.isUpdatingFromRemote = true

            let pasteboard = NSPasteboard.general
            pasteboard.clearContents()
            pasteboard.setString(text, forType: .string)

            self.lastKnownContent = text
            self.lastKnownChangeCount = pasteboard.changeCount
            self.lastSyncedContent = text
            self.lastSyncTime = Date(timeIntervalSince1970: timestamp / 1000)

            // Show notification
            self.showClipboardNotification(text: text)

            print("ClipboardSyncService: Updated local clipboard from Android")

            // Reset flag after delay
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self.isUpdatingFromRemote = false
            }
        }
    }

    /// Show notification when clipboard is updated from phone
    private func showClipboardNotification(text: String) {
        let notification = NSUserNotification()
        notification.title = "Clipboard Synced"
        notification.informativeText = text.count > 50 ? String(text.prefix(50)) + "..." : text
        notification.soundName = nil

        NSUserNotificationCenter.default.deliver(notification)
    }

    /// Manually sync current clipboard
    func syncNow() {
        guard let text = NSPasteboard.general.string(forType: .string),
              !text.isEmpty,
              text.count <= maxClipboardLength else { return }

        syncToFirebase(text: text)
    }

    /// Get current clipboard content
    func getCurrentClipboard() -> String? {
        return NSPasteboard.general.string(forType: .string)
    }

    /// Reduce sync frequency for battery saving
    func reduceFrequency() {
        // Increase check interval from 1.0 to 5.0 seconds
        stopMonitoringLocalClipboard()
        if isEnabled && currentUserId != nil {
            startMonitoringLocalClipboardWithInterval(5.0)
        }
    }

    /// Pause clipboard sync temporarily
    func pauseSync() {
        isEnabled = false
        stopMonitoringLocalClipboard()
        print("ClipboardSyncService: Paused syncing")
    }

    /// Resume clipboard sync
    func resumeSync() {
        isEnabled = true
        if let userId = currentUserId {
            startSync(userId: userId)
        }
    }

    /// Start monitoring with custom interval
    private func startMonitoringLocalClipboardWithInterval(_ interval: TimeInterval) {
        DispatchQueue.main.async { [weak self] in
            self?.clipboardCheckTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
                self?.checkLocalClipboard()
            }
            if let timer = self?.clipboardCheckTimer {
                RunLoop.main.add(timer, forMode: .common)
            }
            print("ClipboardSyncService: Local clipboard monitoring started with \(interval)s interval")
        }
    }
}

File: app/src/main/java/com/phoneintegration/app/desktop/ClipboardSyncService.kt
package com.phoneintegration.app.desktop

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import kotlinx.coroutines.*
import kotlinx.coroutines.tasks.await

/**
 * Service that syncs clipboard content between Android and macOS/desktop.
 * Supports text content syncing in real-time.
 */
class ClipboardSyncService(context: Context) {
    private val appContext: Context = context.applicationContext
    private val auth = FirebaseAuth.getInstance()
    private val database = FirebaseDatabase.getInstance()

    private val clipboardManager = appContext.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    private var clipboardListener: ClipboardManager.OnPrimaryClipChangedListener? = null
    private var clipboardValueListener: ValueEventListener? = null

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // Track last synced content to avoid loops
    private var lastSyncedContent: String? = null
    private var lastSyncedTimestamp: Long = 0
    private var isUpdatingFromRemote = false

    companion object {
        private const val TAG = "ClipboardSyncService"
        private const val CLIPBOARD_PATH = "clipboard"
        private const val USERS_PATH = "users"
        private const val MAX_CLIPBOARD_LENGTH = 50000 // 50KB max for text
        private const val SYNC_DEBOUNCE_MS = 500L
    }

    /**
     * Data class representing clipboard content
     */
    data class ClipboardContent(
        val text: String,
        val timestamp: Long,
        val source: String, // "android" or "macos"
        val type: String = "text" // Currently only "text" supported
    )

    /**
     * Start clipboard sync - monitors local clipboard and listens for remote changes
     */
    fun startSync() {
        Log.d(TAG, "Starting clipboard sync")
        database.goOnline()
        stopListeningForLocalClipboard()
        stopListeningForRemoteClipboard()
        startListeningForLocalClipboard()
        startListeningForRemoteClipboard()
    }

    /**
     * Stop clipboard sync
     */
    fun stopSync() {
        Log.d(TAG, "Stopping clipboard sync")
        stopListeningForLocalClipboard()
        stopListeningForRemoteClipboard()
        scope.cancel()
    }

    /**
     * Register listener for local clipboard changes
     */
    private fun startListeningForLocalClipboard() {
        stopListeningForLocalClipboard()
        clipboardListener = ClipboardManager.OnPrimaryClipChangedListener {
            // Ignore if we're currently updating from remote
            if (isUpdatingFromRemote) {
                Log.d(TAG, "Ignoring clipboard change - updating from remote")
                return@OnPrimaryClipChangedListener
            }

            val clipData = clipboardManager.primaryClip
            if (clipData == null || clipData.itemCount == 0) return@OnPrimaryClipChangedListener

            val item = clipData.getItemAt(0)
            val text = item.text?.toString()

            if (text.isNullOrBlank()) return@OnPrimaryClipChangedListener
            if (text.length > MAX_CLIPBOARD_LENGTH) {
                Log.w(TAG, "Clipboard content too large (${text.length} chars), skipping sync")
                return@OnPrimaryClipChangedListener
            }

            // Debounce and check if content changed
            val now = System.currentTimeMillis()
            if (text == lastSyncedContent && now - lastSyncedTimestamp < 2000) {
                Log.d(TAG, "Clipboard content unchanged, skipping sync")
                return@OnPrimaryClipChangedListener
            }

            scope.launch {
                delay(SYNC_DEBOUNCE_MS) // Debounce
                syncClipboardToFirebase(text)
            }
        }

        clipboardManager.addPrimaryClipChangedListener(clipboardListener)
        Log.d(TAG, "Local clipboard listener registered")
    }

    private fun stopListeningForLocalClipboard() {
        clipboardListener?.let {
            clipboardManager.removePrimaryClipChangedListener(it)
        }
        clipboardListener = null
    }

    /**
     * Listen for clipboard changes from other devices
     */
    private fun startListeningForRemoteClipboard() {
        scope.launch {
            try {
                val currentUser = auth.currentUser ?: return@launch
                val userId = currentUser.uid

                val clipboardRef = database.reference
                    .child(USERS_PATH)
                    .child(userId)
                    .child(CLIPBOARD_PATH)

                val listener = object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        val source = snapshot.child("source").value as? String ?: return

                        // Only process if from another device
                        if (source == "android") return

                        val text = snapshot.child("text").value as? String ?: return
                        val timestamp = snapshot.child("timestamp").value as? Long ?: return

                        // Check if this is newer than what we have
                        if (timestamp <= lastSyncedTimestamp) return

                        Log.d(TAG, "Received clipboard from $source: ${text.take(50)}...")

                        // Update local clipboard
                        updateLocalClipboard(text, timestamp)
                    }

                    override fun onCancelled(error: DatabaseError) {
                        Log.e(TAG, "Clipboard listener cancelled: ${error.message}")
                    }
                }

                clipboardValueListener = listener
                clipboardRef.addValueEventListener(listener)

                Log.d(TAG, "Remote clipboard listener registered")
            } catch (e: Exception) {
                Log.e(TAG, "Error starting remote clipboard listener", e)
            }
        }
    }

    private fun stopListeningForRemoteClipboard() {
        clipboardValueListener?.let { listener ->
            scope.launch {
                try {
                    val userId = auth.currentUser?.uid ?: return@launch
                    database.reference
                        .child(USERS_PATH)
                        .child(userId)
                        .child(CLIPBOARD_PATH)
                        .removeEventListener(listener)
                } catch (e: Exception) {
                    Log.e(TAG, "Error removing clipboard listener", e)
                }
            }
        }
        clipboardValueListener = null
    }

    /**
     * Sync local clipboard content to Firebase
     */
    private suspend fun syncClipboardToFirebase(text: String) {
        try {
            val currentUser = auth.currentUser ?: return
            val userId = currentUser.uid

            val clipboardRef = database.reference
                .child(USERS_PATH)
                .child(userId)
                .child(CLIPBOARD_PATH)

            val timestamp = System.currentTimeMillis()

            val clipboardData = mapOf(
                "text" to text,
                "timestamp" to timestamp,
                "source" to "android",
                "type" to "text"
            )

            clipboardRef.setValue(clipboardData).await()

            lastSyncedContent = text
            lastSyncedTimestamp = timestamp

            Log.d(TAG, "Clipboard synced to Firebase: ${text.take(50)}...")
        } catch (e: Exception) {
            Log.e(TAG, "Error syncing clipboard", e)
        }
    }

    /**
     * Update local clipboard with remote content
     */
    private fun updateLocalClipboard(text: String, timestamp: Long) {
        try {
            isUpdatingFromRemote = true

            val clip = ClipData.newPlainText("SyncFlow", text)
            clipboardManager.setPrimaryClip(clip)

            lastSyncedContent = text
            lastSyncedTimestamp = timestamp

            Log.d(TAG, "Local clipboard updated from remote")

            // Reset flag after a short delay
            scope.launch {
                delay(500)
                isUpdatingFromRemote = false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error updating local clipboard", e)
            isUpdatingFromRemote = false
        }
    }

    /**
     * Manually sync current clipboard
     */
    suspend fun syncNow() {
        val clipData = clipboardManager.primaryClip
        if (clipData == null || clipData.itemCount == 0) return

        val text = clipData.getItemAt(0).text?.toString()
        if (!text.isNullOrBlank() && text.length <= MAX_CLIPBOARD_LENGTH) {
            syncClipboardToFirebase(text)
        }
    }

    /**
     * Get current clipboard content
     */
    fun getCurrentClipboard(): String? {
        val clipData = clipboardManager.primaryClip
        if (clipData == null || clipData.itemCount == 0) return null
        return clipData.getItemAt(0).text?.toString()
    }
}
